virt_lightning/api.py

```diff
@@ -0,0 +1,442 @@
+
+
+def start(configuration, context="default", enable_console=False, **kwargs):
+    print(enable_console)
+    """
+    Start a single VM
+    """
+    conn = libvirt.open(configuration.libvirt_uri)
+    hv = vl.LibvirtHypervisor(conn)
+    hv.init_network(configuration.network_name, configuration.network_cidr)
+    hv.init_storage_pool(configuration.storage_pool)
+    host = {
+        k: kwargs[k] for k in ["name", "distro", "memory", "vcpus"] if kwargs.get(k)
+    }
+    domain = _start_domain(hv, host, context, configuration)
+    if not domain:
+        return
+
+    loop = asyncio.get_event_loop()
+
+    if enable_console:
+        import time
+
+        time.sleep(4)
+        stream = conn.newStream(libvirt.VIR_STREAM_NONBLOCK)
+        console = domain.dom.openConsole(None, stream, 0)
+
+        loop = kwargs.get("loop") or asyncio.get_event_loop()
+        _register_aio_virt_impl(loop)
+
+        def stream_callback(stream, events, _):
+            content = stream.recv(1024).decode()
+            sys.stdout.write(content)
+
+        stream.eventAddCallback(
+            libvirt.VIR_STREAM_EVENT_READABLE, stream_callback, console
+        )
+
+    async def deploy():
+        await domain.reachable()
+
+    loop.run_until_complete(deploy())
+    logger.info(  # noqa: T001
+        (
+            "\033[0m\n**** System is online ****\n"
+            "To connect use:\n"
+            "  vl console %s (virsh console)"
+            "  vl ssh %s"
+        ),
+        domain.name,
+        domain.name,
+    )
+    if kwargs.get("ssh"):
+        domain.exec_ssh()
+
+
+def stop(configuration, **kwargs):
+    """
+    Stop a given VM
+    """
+    conn = libvirt.open(configuration.libvirt_uri)
+    hv = vl.LibvirtHypervisor(conn)
+    hv.init_network(configuration.network_name, configuration.network_cidr)
+    hv.init_storage_pool(configuration.storage_pool)
+    domain = hv.get_domain_by_name(kwargs["name"])
+    if not domain:
+        vm_list = [d.name for d in hv.list_domains()]
+        if vm_list:
+            logger.info("No VM called %s in: %s", kwargs["name"], ", ".join(vm_list))
+        else:
+            logger.info("No running VM.")
+        exit(1)
+    hv.clean_up(domain)
+
+
+def ansible_inventory(configuration, context="default", **kwargs):
+    """
+    Generate an Ansible inventory based on the running VM
+    """
+    conn = libvirt.open(configuration.libvirt_uri)
+    hv = vl.LibvirtHypervisor(conn)
+
+    ssh_cmd_template = (
+        "{name} ansible_host={ipv4} ansible_user={username} "
+        "ansible_python_interpreter={python_interpreter} "
+        'ansible_ssh_common_args="-o UserKnownHostsFile=/dev/null '
+        '-o StrictHostKeyChecking=no"\n'
+    )
+
+    output = ""
+    groups = {}
+    for domain in hv.list_domains():
+        if domain.context != context:
+            continue
+
+        for group in domain.groups:
+            if group not in groups:
```

Comment:
I think this line (and some similar here and there) might be eliminated by using `collections.defaultdict` with `list` as a default factory, just like in a first example from here : https://docs.python.org/3/library/collections.html#defaultdict-examples

like : 

```
import collections
...
groups = collections.defaultdict(list)

...
    for group in domains.groups:
         groups[group].append(domain)
```

`not in` check is not needed with `defaultdict`, and `.items()` work the same as in regular dictionary

Reply:
Awesome, thanks @mksh :-)